contracts/cache_backend.go

```diff
@@ -0,0 +1,114 @@
+package contracts
+
+import (
+	"math/big"
+
+	"github.com/celo-org/celo-blockchain/common"
+	"github.com/celo-org/celo-blockchain/core/vm"
+	"github.com/celo-org/celo-blockchain/metrics"
+	lru "github.com/hashicorp/golang-lru"
+)
+
+var (
+	cacheHits    = metrics.NewRegisteredMeter("contract_comm/caller/cache_hits", nil)
+	cacheMisses  = metrics.NewRegisteredMeter("contract_comm/caller/cache_misses", nil)
+	cacheBadCast = metrics.NewRegisteredMeter("contract_comm/caller/cache_bad_cast", nil)
+)
+
+type cacheKey struct {
+	stateRoot common.Hash
+	caller    vm.ContractRef
+	addr      common.Address
+	input     []byte
```

Comment:
Needs to be a non-slice type: https://play.golang.org/p/0vSKQnveUiW

Reply:
I'm likewise wondering about `caller`, above, since it's an interface.  To be safe, it would have to be something like `callerAddr common.Address` and to get it we'd call `caller.Address()`?